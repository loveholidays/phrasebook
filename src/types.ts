export interface TranslationData {
  [key: string]: string | TranslationData;
}

export type Namespaces = Record<string, TranslationData>;

export type TranslationArgumentValue = string | number | undefined;

export type TranslationArguments = {
  count?: number;
  context?: string;
  numberLocale?: Locale;
  [name: string]: TranslationArgumentValue;
};

export type TFunction = (key: string, args?: TranslationArguments) => string;

export type Locale = string;
