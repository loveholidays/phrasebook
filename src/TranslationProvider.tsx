import React, { createContext, useContext, useMemo } from 'react';
import { DEFAULT_NAMESPACE } from './constants';

import { processTranslation } from './processTranslation';
import type {
  Namespaces, Locale, TFunction, TranslationData,
} from './types';

export interface TranslationContextValue {
  locale: Locale;
  namespaces: Namespaces;
  t: TFunction;
}

export const TranslationContext = createContext<TranslationContextValue>({
  locale: '',
  namespaces: {},
  t: () => '',
});

export interface ReplaceArgumentErrorParams {
  key: string;
  argumentName: string;
}

export type OnErrorCallback =
  & ((errorType: 'REPLACE_ARGUMENT_NOT_PASSED', params: ReplaceArgumentErrorParams) => void);

interface TranslationProviderProps {
  locale: Locale;
  namespaces?: Namespaces;
  translations?: TranslationData;
  onError?: OnErrorCallback;
}

// @TODO:
// - option for deep merge?
// - option for providing a custom `mergeNamespaces` function?
const mergeNamespaces = (
  parentNamespaces: Namespaces = {},
  namespaces: Namespaces = {},
  translations?: TranslationData,
): Namespaces => ({
  ...parentNamespaces,
  ...namespaces,
  ...(translations && {
    [DEFAULT_NAMESPACE]: translations,
  }),
});

export interface UseTranslationReturnValue {
  t: TFunction;
}

export const useTranslation = (namespace?: string) => ({
  t: namespace
    ? (key, args) => useContext(TranslationContext).t(key, {
      ns: namespace,
      ...args,
    })
    : useContext(TranslationContext).t,
}) as UseTranslationReturnValue;

/**
 * TranslationProvider
 * Provider used to create the localisation context.
 * @param {object} props
 * @param {string} props.locale String is used for locale specific formatting.
 * @param {object} props.namespaces Namespaced translations where the keys are the names of the namespaces and the values are the translations for the given namespace.
 * @param {object} props.translations Translation for default locale.
 * @param {function} props.onError Callback could be used to track the error during translation processing.
 * */
export const TranslationProvider: React.FC<React.PropsWithChildren<TranslationProviderProps>> = ({
  locale,
  namespaces,
  translations,
  onError,
  children,
}) => {
  const { namespaces: parentNamespaces } = useContext(TranslationContext);
  const mergedNamespaces = useMemo(
    () => mergeNamespaces(parentNamespaces, namespaces, translations),
    [ parentNamespaces, namespaces, translations ],
  );

  return (
    <TranslationContext.Provider
      value={{
        locale,
        namespaces: mergedNamespaces,
        t: (key, args) => processTranslation({
          locale,
          namespaces: mergedNamespaces,
          key,
          args,
          onError,
        }),
      }}
    >
      {children}
    </TranslationContext.Provider>
  );
};

