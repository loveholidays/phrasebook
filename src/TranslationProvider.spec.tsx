import React from 'react';

import { render } from '@testing-library/react';
import { TranslationProvider, useTranslation } from './TranslationProvider';
import { TranslationArguments } from './types';

import namespaces from './testing/testNamespaces.json';

interface TestComponentProps {
  name: string;
  namespace?: string;
  args?: TranslationArguments;
}

const TestComponent: React.FC<TestComponentProps> = ({
  name,
  namespace,
  args,
}) => {
  const { t } = useTranslation(namespace);

  return (
    <span>{t(name, args)}</span>
  );
};

describe('TranslationProvider', () => {
  const locale = 'en-GB';

  it('exposes the translation function via useTranslation hook', () => {
    const { asFragment } = render(
      <TranslationProvider
        locale={locale}
        namespaces={namespaces}
      >
        <TestComponent
          name="search.label"
        />
        <TestComponent
          name="reviews"
          args={{ count: 1 }}
        />
        <TestComponent
          name="reviews"
          args={{ count: 4 }}
        />
        <TestComponent
          name="boardBasis.code"
        />
        <TestComponent
          name="boardBasis.code"
          args={{ context: 'AI' }}
        />
        <TestComponent
          name="title"
          namespace="ns1"
        />
        <TestComponent
          name="title"
          args={{ ns: 'ns2' }}
        />
      </TranslationProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('merges namespaces when nesting providers', () => {
    const { asFragment } = render(
      <TranslationProvider
        locale={locale}
        namespaces={{
          ns1: namespaces.ns1,
        }}
      >
        <TranslationProvider
          locale={locale}
          namespaces={{
            ns2: namespaces.ns2,
          }}
        >
          <TestComponent
            name="title"
            namespace="ns1"
          />
          <TestComponent
            name="title"
            args={{ ns: 'ns2' }}
          />
        </TranslationProvider>
      </TranslationProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should override number format', () => {
    const { asFragment } = render(
      <TranslationProvider
        locale={locale}
        namespaces={namespaces}
      >
        <TestComponent
          name="reviews"
          args={{ count: 12345 }}
        />
        <TestComponent
          name="reviews"
          args={{ count: 12345, overrideLocale: 'de-DE' }}
        />
      </TranslationProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
