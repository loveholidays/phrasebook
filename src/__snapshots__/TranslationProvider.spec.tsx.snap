// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TranslationProvider exposes the translation function via useTranslation hook 1`] = `
<DocumentFragment>
  <span>
    Search
  </span>
  <span>
    1 review
  </span>
  <span>
    4 reviews
  </span>
  <span>
    Board Basis
  </span>
  <span>
    All Inclusive hotel
  </span>
  <span>
    title from namespace 1
  </span>
  <span>
    title from namespace 2
  </span>
</DocumentFragment>
`;

exports[`TranslationProvider merges namespaces when nesting providers 1`] = `
<DocumentFragment>
  <span>
    title from namespace 1
  </span>
  <span>
    title from namespace 2
  </span>
</DocumentFragment>
`;

exports[`TranslationProvider should override number format 1`] = `
<DocumentFragment>
  <span>
    12,345 reviews
  </span>
  <span>
    12.345 reviews
  </span>
</DocumentFragment>
`;
